docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+For example, the Job object could have a list of PodSpecs to run.
+However, this is so general that it provides little value.  It would:
+
+- make the Job Spec very verbose, especially for jobs with thousands of work items
+- Job becomes such a vague concept that it is hard to explain to users
+- in practice, we do not see cases where many pods which differ across many fields of their
+  specs, and need to run as a group, with no ordering constraints.
+- CLIs and UIs need to support more options for creating Job
+- it is useful for monitoring and accounting databases want to aggregate data for pods
+  with the same controller.  However, pods with very different Specs may not make sense
+  to aggregate.
+- profiling, debugging, accounting, auditing and monitoring tools cannot assume common
+  images/files, behaviors, provenance and so on between Pods of a Job.
+
+Also, variety has another cost.  Pods which differ in ways that affect scheduling
+(node constraints, resource requirements, labels) prevent the scheduler
+from treating them as fungible, which is an important optimization for the scheduler.
+
+Therefore, we will not allow Pods from the same Job to differ arbitrarily
+(anyway, users can use multiple Job objects for that case).  We will try to
+allow as little as possible to differ between pods of the same Job, while
+still allowing users to express common parallel patterns easily For users
+who need to run jobs which differ in other ways, they can create multiple
+Jobs, and manage them as a group using labels.
+
+
+From the above work lists, we see a need for Pods which differ in their command
+lines, and in their environment variables.  These work lists do not require the
+pods to differ in other ways.
+
+Experience in a [similar systems](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43438.pdf) has shown this model to be applicable
+to a very broad range of problems, despite this restriction.
+
+Therefore we will allow pods in the same Job to differ **only** in the following aspects:
+
+- command line
+- environment variables
+
+That is a statements about what the Job resource schema can express, and what
+the Job controller will do when creating pods of the same job.
+
+An admission controller could still modify pods, and some fields of pods could still
+be updated after creation.  Nothing will prevent this.
+Pods of the same Job will not have different volumes mounted on them, though
+users can mount all needed volumes and then select different files of interest.
+
+#### Interaction with vertical autoscaling
+
+Since pods of the same job will not be created with different resources,
+a vertical autoscaler will need to:
+
+- if it has index-specific initial resource suggestions, suggest those at admission
+time; it will need to understand indexes.
+- mutate resource requests on already created pods based on usage trend or previous container failures
+- modify the job template, affecting all indexes.
```

Comment:
Still it sounds weird for me, that the vertical scaling would change in any way the podspec I've created, as a user. For me vertical scaling is the interest of a cluster administrator, rather than for an end user submitting jobs, isn't it?